rtio: Asynchronous Real-Time I/O

teburd opened the pull request:
Enabling the usage of asynchronous I/O with interrupt or DMA driven transfers across multiple devices with an io_uring like API.

The docs are the most helpful starting place
https://github.com/teburd/zephyr/blob/rtio_next/doc/services/rtio/index.rst

Enhancements moved to Issue https://github.com/zephyrproject-rtos/zephyr/issues/46658

This is an experimental API and subsystem and it is useful now but will require improvements and refinements to cover the expected use cases.

Commit:
rtio: Real-Time Input/Output Stream

A DMA friendly Stream API for zephyr. Based on ideas from io_uring
and iio, a queue based API for I/O operations.

Provides a pair of fixed length ringbuffer backed queues for submitting
I/O requests and recieving I/O completions. The requests may be chained
together to ensure the next operation does not start until the current
one is complete.

Requests target an abstract rtio_iodev which is expected to wrap all
the hardware particulars of how to perform the operation. For example
with a SPI bus device, a description of what a read, and write mean
can be decided by the iodev wrapping a particular device
hanging off of a SPI controller.

The queue pair are submitted to an executor which may be a simple
inplace looping executor done in the callers execution context
(thread/stack) but other executors are expected. A threadpool executor
might for example allow for concurrent request chains to execute in
parallel. A DMA executor, in conjunction with DMA aware iodevs
would allow for hardware offloading of operations going so far as to
schedule with priority using hardware arbitration.

Both the iodev and executor are definable by a particular
SoC, meaning they can work in conjuction to perform IO operations
using a particular DMA controller or methodology if desired.

The application decides entirely how large the queues are, where
the buffers to read/write come from (some executors
may have particular demands!), and which executor to submit
requests to.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

brockus-zephyr commented on doc/services/rtio/index.rst:
```diff
@@ -0,0 +1,417 @@
+Canceling an already queued operation could be possible with a small
+API addition to perhaps take both the RTIO context and a pointer to the
+submission queue entry. However, cancellation as an API induces many potential
+complexities that might not be appropriate. It's something to be decided upon.
```
Canceling may make sense at least for times or failures and timeouts. This would free the resource of wasting time in case another SQE is waiting to run on that device

teburd replied:
Yes, though has some potential implications. Right now the sqe is setup in a pair of acquire/produce calls which wrap the safe area in which to do mutation of a sqe from the application. By introducing cancel, there needs to be a way to continue mutating the sqe in some safe form, likely with an atomic bit flag that gets reset on acquire again (so its a best effort cancel, not a blocking locking cancel). There's also some memory usage additions from cancel then for each SQE. An ifdef would seem appropriate as not everyone is likely to care about it.

brockus-zephyr replied:
I think best effort is better than possibly hanging up a device for no reason. I am always good with ifdef reducing footprint on unneeded features.

brockus-zephyr commented on doc/services/rtio/index.rst:
```diff
@@ -0,0 +1,417 @@
+=========================
+
+.. code-block:: C
+
```
Adding comments through out the examples to make reading easier on the reader would be advised.  Assume your reader knows how to code but has no idea what any of these functions are doing. Adding a general concept on each block would make all the difference in the world.

teburd replied:
I agree, I will go back through these samples and update them to the current API with commentary

brockus-zephyr commented on doc/services/rtio/index.rst:
```diff
@@ -0,0 +1,417 @@
+		const struct rtio_executor *rtio_dma_exec =
+				dma_rtio_executor(dma);
+
+        /*
```
nit: this is driving me nuts... one more tab please

MaureenHelm committed:
samples: rtio: Add sensor batch processing sample application

Adds a new sample application that demonstrates using the RTIO subsystem
to read periodic sensor data directly into buffers allocated by the
application, asynchronously process batches of data with an algorithm,
and recycle buffers back for reading additional sensor data.

The sensor iodev in this application is an timer-driven device that
executes one read request per timer period. It doesn't actually send any
transactions to a real I2C/SPI bus or read any real data into the
application-provided buffers. This timer-driven behavior mimics how a
real sensor periodically triggers a GPIO interrupt when new data is
ready.

The sensor iodev currently uses an internal message queue to store
pending requests from the time they are submitted until the next timer
expiration. At least one pending request needs to be stored by the iodev
to ensure that it has a buffer available to read data into. However,
any more than that should probably be handled by the application, since
it's the application that determines how often it can submit new
requests and therefore how deep the queue needs to be.

The sensor iodev is implemented to support multiple instances with
devicetree, but additional work remains to enable and use more than one
in the application.

Tested on native_posix and frdm_k64f.

Signed-off-by: Maureen Helm <maureen.helm@intel.com>

teburd commented on doc/services/rtio/index.rst:
```diff
@@ -0,0 +1,417 @@
+In order to know when a sqe has completed there is a completion
+queue (cq) with completion queue events (cqe). A sqe once completed results in
+a cqe being pushed into the cq. The ordering of cqe may not be the same order of
+sqe. A chain of sqe will however ensure ordering and failure cascading.
```
The ordering may not be the same. The cqe and sqe are tied together with the common void *userdata which is passed along. While this userdata could point at something like a struct or buffer, it could also directly store an integer identifier like a sequential index, that's up to the user of the API.

brockus-zephyr replied:
So it is left to the originator to determine which CQE it has just consumed, correct?

teburd replied:
Yes it is, that's correct.

brockus-zephyr replied:
It sort of seems like adding the offset index of the entry that is being sent to the executor would be easy enough and would allow the originator less grief finding what has and has not been executed in a common orderly manner.

teburd commented on doc/services/rtio/index.rst:
```diff
@@ -0,0 +1,417 @@
+
+Multiple operations against the same iodev are done in the order provided as
+soon as possible. If two operation chains have varying points using the same
+device its possible one chain will have to wait for another to complete.
```
Consider wanting to read data from two devices on the same bus, and I believe the answer to this is potentially yes. That has some implications that I've been mulling over.

brockus-zephyr replied:
For some higher level devices, I can understand a yes on this... If you have something like an I2C that one chain is doing a START/WRITE/READ/READ/RESTART/WRITE/STOP. That would not really map with the restart but otherwise would be a Connect/TX/RX/RX/????/TX/Disconnect. If it were required that this be synchronous each phase of this would happen one step at a time. If there were also another chain that was good to go that were talking to that same device it seems like this could interfer with the current thread that sort of owns the device until Disconnect...
Also the Disconnect would be locked out if there were a failure on one of the previous operations, say a timeout. Would that possibly lock out other chains or at least leave the bus in an odd state?

teburd replied:
Yeah this is all correct, the implications are that other chains would need to, as part of the request, enqueue themselves to be notified when the iodev is available. Maybe this can all be implied in a connect type request, where if chain 1 is connected and chain 2 then requests to connect, that chain 2 connect sqe won't complete until chain 1 fails or disconnects. A failure on a chain should implicitly disconnect in some manner, probably by calling an iodev provided API like chain_failed(), and the iodev can decide what to do if its in a transactional state.

teburd commented on doc/services/rtio/index.rst:
```diff
@@ -0,0 +1,417 @@
+Because the chip select and clocking is common to a particular SPI controller
+and device on the bus it is embodied in what RTIO calls an iodev.
+
+Multiple operations against the same iodev are done in the order provided as
```
Per bus controller, equivalent to the struct device's available today

teburd commented on doc/services/rtio/index.rst:
```diff
@@ -0,0 +1,417 @@
+
+If anything in this chain of operations fails give up. Some of those operations
+can be embodied in a device abstraction that understand a read or write
+implicitly means setup the clock and chip select. The transactional nature of
```
It probably makes sense to make this more explicit, akin to i2c start/stop or spi chip select. There are some hardware IPs that enable ISR driven state machines around even these steps. For example the GigaDevice i2c peripheral. I like the connect/disconnect terminology as its fairly universal and implies there's some setup and teardown time involved which is true!

teburd commented:
@brockus-zephyr thank you for the very insightful comments along with grammatical and formatting corrections, those are all very welcome

teburd commented on subsys/rtio/rtio_executor_inplace.c:
```diff
@@ -0,0 +1,39 @@
+		 * Per the rtio_iodev_submit docs, with SYNC is flagged the call
+		 * MUST block until the submission is complete in some way.
+		 */
+		bool blocking = (completions < wait_count) || (sqe->flags & RTIO_SQE_CHAINED);
```
That of course makes sense, this was simple to implement and test out. I'll say this is one of the many reasons this is an RFC and not a ready to go PR :racing_car: 

I'm working on an improvement in the way things are going to wait on completions either by a semaphore or some other method of sleeping the calling thread (if a thread) until the wait_count is completed

brockus-zephyr commented on include/zephyr/rtio/rtio.h:
```diff
@@ -0,0 +1,418 @@
+ * @brief Submit a request to be done while blocking the caller
+ *
+ * The submitted operation will be done synchronously and the call shall
+ * not return until the operation is complete. The default is a non-blocking
```
"until the operation is complete"? Is this an execution wall with all pending async operations needing to be complete to fulfill this requirement or does this just the queue will fly again when that last entry completes?

teburd replied:
This flag is going to go away in the next round of my work as it will be unnecessary

brockus-zephyr commented on include/zephyr/rtio/rtio.h:
```diff
@@ -0,0 +1,418 @@
+{
+	sqe->op = RTIO_OP_NOP;
+	sqe->iodev = iodev;
+	sqe->userdata = userdata;
```
why is userdata only represented in the do nothing case?

teburd replied:
A mistake that is being corrected in another round of work

brockus-zephyr commented on include/zephyr/rtio/rtio.h:
```diff
@@ -0,0 +1,418 @@
+}
+
+/**
+ * @brief Perform an submitted operation, maybe block or not based on the blocking parameter
```
nit: Perform a submitted operation

brockus-zephyr commented on include/zephyr/rtio/rtio.h:
```diff
@@ -0,0 +1,418 @@
+	 * completion
+	 *
+	 * If unique identification of completions is desired this should be
+	 * unique as well.
```
Why not add the CQ index of the CQE as it is delivered to the executor and then provide that automatically in the SQE when it is placed in the SQ. This would be common and less hassle for the originator.  I am fine with userdata being added as well but don't think this is a easiest way for the originator to determine what status they are looking at.

teburd replied:
Looking closer at how libiouring does this, with a integer rather than a pointer (it could store a pointer) I think this makes a sensible default.

brockus-zephyr commented on include/zephyr/rtio/rtio.h:
```diff
@@ -0,0 +1,418 @@
+ * @brief A completion queue event
+ */
+struct rtio_cqe {
+	int32_t result; /**< Result from operation */
```
that possible other index could live here to align the SQE offset within the originators SQ request to the CQE that has completed.

teburd replied:
Agreed, but perhaps changing void *userdata to uintptr_t userdata and having the default be the ringbuffer index of the sqe makes sense.

brockus-zephyr replied:
I am fine with just using the index instead of a pointer to userdata.

brockus-zephyr commented on include/zephyr/rtio/rtio.h:
```diff
@@ -0,0 +1,418 @@
+	 * @brief Submission function for a request to the iodev
+	 *
+	 * By default with no flags provided the function should
+	 * return as soon as possible. It is acceptable but not
```
By default, with no flags provided, the function should...

brockus-zephyr commented on include/zephyr/rtio/rtio.h:
```diff
@@ -0,0 +1,418 @@
+	 * of the request.
+	 *
+	 * If the RTIO_SUBMIT_SYNC flag is set the function MUST
+	 * complete the request before returning.
```
I keep seeing the same pattern that implies this flag only bottlenecks on this specific request and not any ASYNC requests that also preceeded this request. This could make the idea of ASYNC unusable in some instances.

teburd replied:
This flag is going to go away in the next iteration of this, which should address most of the concerns surrounding it I think. It was there initially to enable the simple executor loop I had created with a wait count that used a blocking call until the wait count was completed. The thinking was you are waiting on the count anyways so may as well block the calling thread until the requested wait count of submissions was complete.

Your commentary has me really thinking this is wrong, as really all operations should be asynchronous and the wait count done in a different method that allows the calling thread to sleep if desired (though polling is still possible)

brockus-zephyr commented on include/zephyr/rtio/rtio.h:
```diff
@@ -0,0 +1,418 @@
+	 * complete the request before returning.
+	 *
+	 * Since the entire context (including the executor) is
+	 * passed in the executor may be used in conjunction with the
```
is passed in, the executor

brockus-zephyr commented on include/zephyr/rtio/rtio.h:
```diff
@@ -0,0 +1,418 @@
+}
+
+/**
+ * @brief Acquire and prepare a read request
```
where is the aquire portion of this?

teburd replied:
Originally this acquired from the sq and prepped, I've since corrected this and the other similar doc strings in an updated commit

brockus-zephyr commented on include/zephyr/rtio/rtio.h:
```diff
@@ -0,0 +1,418 @@
+}
+
+/**
+ * @brief Acquire and prepare a write request
```
where is the aquire portion of this?

teburd replied:
ditto as above

brockus-zephyr commented on include/zephyr/rtio/rtio_executor_inplace.h:
```diff
@@ -0,0 +1,82 @@
+ * as needed. Different operation chains are done in order and are not done
+ * in parallel using threads or DMA. Each rtio_iodev may still be asynchronous
+ * and perhaps even use the platforms DMA, but no specifics are detailed here
+ * other than the device is asked to perform an IO operation with a buffer.
```
Is priority based ordering used in this or just simple FIFO?

teburd replied:
priority is given and used when DMA is to be used. Potentially it might be used in the future in other ways, but predominantly I've added because DMA controllers typically let you set priorities that get used by the hardware to prioritize things.

brockus-zephyr commented on include/zephyr/rtio/rtio_executor_inplace.h:
```diff
@@ -0,0 +1,82 @@
+ * as needed. Different operation chains are done in order and are not done
+ * in parallel using threads or DMA. Each rtio_iodev may still be asynchronous
+ * and perhaps even use the platforms DMA, but no specifics are detailed here
+ * other than the device is asked to perform an IO operation with a buffer.
```
Why is the same verbage repeated here as it has already been presented on lines 26-30?

teburd replied:
:man_facepalming:

brockus-zephyr commented on include/zephyr/rtio/rtio_executor_inplace.h:
```diff
@@ -0,0 +1,82 @@
+/**
+ * @brief Define a inplace executor with a given name
+ *
+ * The resulting symbol is typed as struct rtio_executor *
```
typed as "struct rtio_executor *" indicates the entire type and the * was not a part of a comment line that was partly removed.

brockus-zephyr commented on include/zephyr/rtio/rtio_spsc.h:
```diff
@@ -0,0 +1,217 @@
+ *
+ * This SPSC implementation works on an array which wraps using a power of two
+ * size and uses a bit mask to perform a modulus. Atomics are used to allow
+ * single-producer single-consumer safe semantics without locks. Elements are
```
This is the first point in all of the files that I locate SPSC is probably "single-producer single-consumer". It should be made obvious to not make your reader go on a hunt for this.

brockus-zephyr commented on include/zephyr/rtio/rtio_spsc.h:
```diff
@@ -0,0 +1,217 @@
+ * expected to be of a fixed size. The API is type safe as the underlying buffer
+ * is typed and all usage is done through macros.
+ *
+ * A spsc may be declared on a stack or statically and work as intended so long
```
If you are anticipating your reader reading this as the letters S P S C then this should be "An SPSC". If on the other hand you expect them to do an inline replace as they are reading this to single-producer single-consumer, then "A" is the correct starting.  Either case, most acronyms are capatilized.

brockus-zephyr commented on include/zephyr/rtio/rtio_spsc.h:
```diff
@@ -0,0 +1,217 @@
+ * Safe usage would be, where A and B are unique execution contexts:
+ * 1. ISR A producing and a Thread B consuming.
+ * 2. Thread A producing and ISR B consuming.
+ * 3. Thread A producing and Thread B consuming.
```
"ISR A producing and ISR B consuming" is not safe?

teburd replied:
That is safe, and should be added. What would not be sensible would be polling for any long period in an ISR and wouldn't be expected to work anyways on non-SMP systems.

brockus-zephyr commented on include/zephyr/rtio/rtio_spsc.h:
```diff
@@ -0,0 +1,217 @@
+};
+
+/**
+ * @brief Statically initialize a rtio_spsc
```
if this will be read as "initialize a Real Time IO" then 'a' is correct, otherwise if rtio_spsc will be read as R T I O S P S C, then "initialize an rtio_spsc"

brockus-zephyr commented on include/zephyr/rtio/rtio_spsc.h:
```diff
@@ -0,0 +1,217 @@
+	}
+
+/**
+ * @brief Declare an anonymous struct type for a rtio_spsc
```
same rtio_spsc comment as above for the preceeding "a"

brockus-zephyr commented on include/zephyr/rtio/rtio_spsc.h:
```diff
@@ -0,0 +1,217 @@
+	} _ ## name
+
+/**
+ * @brief Define a rtio_spsc with a fixed size
```
same rtio_spsc comment as above for the preceeding "a"

brockus-zephyr commented on include/zephyr/rtio/rtio_spsc.h:
```diff
@@ -0,0 +1,217 @@
+	})
+
+/**
+ * @brief Consume a element from the spsc
```
Consume an element

teburd commented on include/zephyr/rtio/rtio.h:
```diff
@@ -0,0 +1,418 @@
+
+	uint16_t flags; /**< Op Flags */
+
+	struct rtio_iodev *iodev; /**< Device to operation on */
```
"Device to submit operation to" might be the most accurate here but yes, the doc string here isn't quite right

teburd commented on include/zephyr/rtio/rtio.h:
```diff
@@ -0,0 +1,418 @@
+ * all chained requests will also fail.
+ *
+ * The completion of these requests are pushed into a fixed size completion
+ * queue which an application may actively poll for completions.
```
Indeed, there are times when you may wish to poll, but most likely you'd like the calling thread to be pending on something and not wasting time. It's something I'm working on now

teburd commented on include/zephyr/rtio/rtio.h:
```diff
@@ -0,0 +1,418 @@
+ *
+ * These operations may be chained in a such a way that only when the current
+ * operation is complete will the next be executed. If the current request fails
+ * all chained requests will also fail.
```
To wait for all pending async operations would be the equivalent of io_uring's flush operation which I don't yet have in mind or feel would be immediately useful.

brockus-zephyr replied:
If I have a couple of devices that are gathering data asynchronously and then going to start using that data... I would need for all of the async data requests to be completed. This is not far fetched and could be quite useful in maybe fused sensor implementation.

teburd replied:
That's a very fair point, in which case this would need something like io_uring's pipe flush

teburd commented on include/zephyr/rtio/rtio.h:
```diff
@@ -0,0 +1,418 @@
+ * @file
+ * @brief Real-Time IO device API for moving bytes with low effort
+ *
+ * RTIO uses a SPSC lock-free queue pair to enable a DMA and ISR friendly I/O API.
```
I had rtio_fifo as the name before, but thought I might want to add a multi-producer single consumer queue for the completion queue given it can potentially accept completions from multiple producers (chains).

I'm open to suggestions on the best route for naming these things. SPSC, MPSC, MPMC are pretty common terms for lock free queues that I've seen though.

brockus-zephyr replied:
I am fine with SPSC as long as it is defined. I spent a lot of years in storage and then some time in networking. The acronyms that I found common in each of these were common within that particular industry but not so much outside. People should feel included when reading this and not like they have to solve a puzzle.  If I Google "DMA computer" I will get Direct Memory Access as the first entry. "ISR computer" gives me the right answer in the first entry.  "SPSC computer" not so much... I have gone pages through the results and get no relevent hits that says that means single-producer single-consumer.

teburd replied:
That is entirely fair and unfortunate its a buried term

teburd commented on doc/services/rtio/index.rst:
```diff
@@ -0,0 +1,417 @@
+*******
+
+An application wishing to do complex DMA or interrupt driven operations today
+in Zephyr requires direct knowledge of the hardware and how it works.
```
Certainly an application needs to know and control the how once the data rates start increasing where efficiency can translate to success or failure. At lower data rates perhaps less interesting but could be a useful knob to tweak with the goal of reducing power usage in exchange for higher latencies perhaps. e.g. tweak the spi clock down, do a dma transfer of a sensor fifo, fall asleep and wake up when done.

teburd commented on doc/services/rtio/index.rst:
```diff
@@ -0,0 +1,417 @@
+
+The above sequence of operations is embodied in RTIO as chain of
+submission queue entries (sqe). Chaining is done by setting a bitflag in
+an sqe to signify the next sqe must wait on the current one.
```
Chaining is meant to say the current SQE MUST complete before the next one is started. The wording here is wrong clearly.

Chains can be thought of as independent threads of execution. You wouldn't expect two C statements to run in parallel, the mental model is one is done the next statement is started.

The submission queue while being a queue itself, is a description of numerous concurrent chains of operations that can be done, ideally driven entirely by hardware and hardware events. But they could be larger operations that map to potentially many hardware events and operations such as a transfer from a sensor or audio codec.

The io_uring guide probably best describes this, direct quote from
[io_uring.pdf](https://kernel.dk/io_uring.pdf)

> The chain of linked sqes continue as long as IOSQE_IO_LINK is set in the flags ﬁeld. Hence the chain is deﬁned as
> starting with the ﬁrst sqe that has IOSQE_IO_LINK set, and ends with the ﬁrst subsequent sqe that does not have it set.
> Arbitrarily long chains are supported.
> The chains execute independently of other sqes in the submission ring. Chains are independent execution units, and
> multiple chains can execute and complete in parallel to each other. This includes sqes that are not part of any chain.

Now is this the right model for RTIO and an RTOS is another question altogether! That's what I'm attempting to answer for myself and so far I think makes sense.

brockus-zephyr replied:
I could see having input from multiple sensors on different busses that send their data to another software sensor that combines them and sends output to the original caller. This could include async operations happening in parallel to the physical sensors but the software sensor may not be able to do anything until all of the data from all of the sensors comes in... Thus putting up a wall to wait for all of the async chains to join back would be a good way to send this data on to the software sensor.

teburd replied:
I can definitely see that! That really is creating a processing pipe. There's potential other flows like that even with something like a soft core on an fpga with a hardware accelerator I can imagine as well, I agree it needs a full pipe wait mode of some kind for that sort of thing.

teburd commented:
I've pushed a rather large update to this.

The executor and the entire asynchronous bits are now significantly clearer I believe. The inplace executor has been renamed to simple as it is a non-concurrent executor. A concurrent one would need to peek into the ringbuffer and understand the memory range of a chain (and thus which executing chain the sqe belongs to). It would also need to *not* consume on each sqe completion, but do some more interesting things to manage freeing back the sqe's to the ringbuffer.

I considered copying these or using a linked list instead, but really... the ringbuffer is just fine, each concurrent executor just needs some extra state.

I have some verbage fixes still that are due given the feedback from @brockus-zephyr and I promise to get to those soon.

teburd commented:
Semaphore and polling waits are now tested and pass on everything, native_posix required a little ifdefry for now to get working with the polling mode.

Ideally polling mode works without a scheduler, and so the polling mode should test for that scenario.

nordic-krch commented:
sorry for delay in review, trouble find time but it's on my todo for next few days.

Commit:
rtio: Low (Memory) Cost Concurrent scheduler

Schedules I/O chains in the same order as they arrive providing a fixed
amount of concurrency. The low memory cost comes at the cost of some
computational cost that is likely to be acceptable with small amounts
of concurrency.

The code cost is about 4x higher than the simple linear executor
which isn't entirely unexpected as the logic requirements are quite a bit
more than doing the next thing in the queue.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

nordic-krch reviewed (commented):
Couple of comments.
In general, i would like to discuss if it's possible to extend rtio to support:
- sequence completion callback
- adding new sqe to active rtio
- mpsc rtio

I have few use cases in mind:
- asynchronous managing of access to i2c/spi device - mpsc rtio would be used for that
- setting up a sequence where you write data to sensor, wait, read from sensor. async notification
some code snippet could look like that (using sem to wait):
```
struct k_sem sem;
struct async_sem asem = { .sem = &sem};
static rtio_cb cb = { .cb = async_semaphore_cb, .userdata = &asem };
rtio_sqe *wr_sqe = rtio_acquire(&rtio);
rtio_sqe *delay_sqe = rtio_acquire(&rtio);
rtio_sqe *read_sqe = rtio_acquire(&rtio);
rtio_write_prep(wr_sqe, rtio_i2c, tx_buf, sizeof(tx_buf));
rtio_delay_prep(delay_seq, &rtio_delay, K_MSEC(10));
rtio_read_prep(rd_sqe, rtio_i2c, rx_buf, sizeof(rx_buf));

err = rtio_submit(&rtio, read_sqe, &cb);

err = k_sem_take(&sem, K_FOREVER);
```

- setting up a sequence where you write data to sensor, wait for gpio change, read from sensor

teburd commented:
> Couple of comments. In general, i would like to discuss if it's possible to extend rtio to support:
> 
> * sequence completion callback

Yeah that would be possible, though intent here is to catch the completions in a thread somewhere, even in the MULTITHREADING=N case I'd expect to want to check the completions in the non-ISR context.


> * adding new sqe to active rtio

io_uring supports this, as should rtio, rtio_submit() should be callable multiple times. Really that just requires checking if there's an ongoing callback chain already proceeding or not or perhaps updating when to stop. io_uring actually has a really nice mode of operation where a thread is constantly running to polling and pulling from the queue, so submit() isn't even needed.

> * mpsc rtio

Managing access to i2c/spi is another level altogether in my mind. It's possible in the same RTIO executor to provide a suspend callback for the iodev (this is my intent) where the chain gets suspended and polled on with other sqe completions for resuming. e.g. two sqe chains attempt to use the same SPI bus, what happens? The first one wins. prio in this case is targetting the DMA channel priority (hardware bus arbitrator) less so the queue of requested I/O operations.

It could use prio, but the code size would probably blow up. It's a strong goal of mine to keep the code size small knowing that usage of this will be limited if its large. At the moment its about 500 bytes or so ROM for the simple scheduler.

> 
> I have few use cases in mind:
> 
> * asynchronous managing of access to i2c/spi device - mpsc rtio would be used for that

Right, this is sort of where things could get complicated and I'm not sure its worth the complications. The same interfaces could be used as they are today but would require a more complex executor that understands there may be multiple contexts submitting. There's almost certainly going to be a code size cost for that and not something everyone need. That's ok though.

The basic thinking for this from me would be each thread still views and has an RTIO context but share a common, multi-queue executor. The common executor can manage I/O priorities globally rather than locally. There's little benefit I feel in sharing the submission/completion queue pair and a lot of complexities in doing so. Does that make sense to you as well?

> * setting up a sequence where you write data to sensor, wait, read from sensor. async notification
>   some code snippet could look like that (using sem to wait):
> 
> ```
> struct k_sem sem;
> struct async_sem asem = { .sem = &sem};
> static rtio_cb cb = { .cb = async_semaphore_cb, .userdata = &asem };
> rtio_sqe *wr_sqe = rtio_acquire(&rtio);
> rtio_sqe *delay_sqe = rtio_acquire(&rtio);
> rtio_sqe *read_sqe = rtio_acquire(&rtio);
> rtio_write_prep(wr_sqe, rtio_i2c, tx_buf, sizeof(tx_buf));
> rtio_delay_prep(delay_seq, &rtio_delay, K_MSEC(10));
> rtio_read_prep(rd_sqe, rtio_i2c, rx_buf, sizeof(rx_buf));
> 
> err = rtio_submit(&rtio, read_sqe, &cb);
> 
> err = k_sem_take(&sem, K_FOREVER);
> ```
> 
> * setting up a sequence where you write data to sensor, wait for gpio change, read from sensor

It's an interesting way to get notified of completion. Maybe the callback could be provided the wait count, or even the CQE to wait for. Something I'll think on. It does seem more flexible than the simple wait count. It's also a nice reuse of the async helpers I PRed.

teburd commented on include/zephyr/rtio/rtio.h:
```diff
@@ -0,0 +1,606 @@
+#ifdef CONFIG_RTIO_SUBMIT_SEM
+
+	if (wait_count > 0) {
+		printk("taking semaphore\n");
```
yeah this printk was a debug friendliness thing for myself along with a few others I need to clean up

teburd commented on include/zephyr/rtio/rtio.h:
```diff
@@ -0,0 +1,606 @@
+	/**
+	 * @brief SQE completes successfully
+	 */
+	void (*ok)(struct rtio *r, const struct rtio_sqe *sqe, int result);
```
Its a question I asked myself, I felt the more definitive and the result is entirely up to the operation this way without any sort of requirements on it.

teburd commented on include/zephyr/rtio/rtio.h:
```diff
@@ -0,0 +1,606 @@
+	 */
+	void *userdata;
+
+	union {
```
Probably more realistically, something like a dmabuf is needed that can understand the dcache/alignment requirements of the platform, at least as one option.

To map to the existing spi/i2c api's it would need to be more like a readv/writev/transcievev (vectored). Maybe that's good enough instead of externalizing those bus transactions.

teburd commented on include/zephyr/rtio/rtio.h:
```diff
@@ -0,0 +1,606 @@
+ * @brief A submission queue event
+ */
+struct rtio_sqe {
+	uint8_t op; /**< Op code */
```
OP is not device specific, its mean to be a generalized set of operations defined here, but could be extended to have iodev specific operations.

The iodev structure could potentially encompass some other common functionality

mbolivar-nordic commented on doc/services/rtio/index.rst:
```diff
@@ -0,0 +1,431 @@
+.. _rtio_api:
+
+Real Time I/O (RTIO)
```
This needs an entry in doc/develop/api/overview.rst as well, likely marking the status as 'experimental'. Please feel free to dismiss my review once you address  this.

Commit:
docs: rtio: Add RTIO API to the API Overview Doc

Notes that the API is currently experimental, and is targetting
being added initially for 3.2.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>
